jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/ExtractTarStep.java

```diff
@@ -146,42 +151,39 @@ public LocalImage call()
 
         for (int index = 0; index < layerFiles.size(); index++) {
           Path file = destination.resolve(layerFiles.get(index));
+          DescriptorDigest diffId = configurationTemplate.getLayerDiffId(index);
 
-          // Compress layers if necessary and calculate the digest/size
-          Blob blob = Blobs.from(file);
           try (ProgressEventDispatcher childDispatcher =
                   childProgressFactories
                       .get(index)
                       .create("compressing " + file, Files.size(file));
               ThrottledAccumulatingConsumer throttledProgressReporter =
                   new ThrottledAccumulatingConsumer(childDispatcher::dispatchProgress)) {
-            if (!layersAreCompressed) {
-              Path compressedFile = destination.resolve(layerFiles.get(index) + ".compressed");
-              try (GZIPOutputStream compressorStream =
-                      new GZIPOutputStream(Files.newOutputStream(compressedFile));
-                  NotifyingOutputStream notifyingOutputStream =
-                      new NotifyingOutputStream(compressorStream, throttledProgressReporter)) {
-                blob.writeTo(notifyingOutputStream);
+            Optional<CachedLayer> optionalLayer = cache.retrieveTarLayer(diffId);
+            if (optionalLayer.isPresent()) {
+              // Retrieve pre-compressed layer from cache
+              CachedLayer layer = optionalLayer.get();
+              layers.add(new PreparedLayer.Builder(layer).build());
+              v22Manifest.addLayer(layer.getSize(), layer.getDigest());
+            } else {
+              // Compress layers and calculate the digest/size
+              Blob blob = Blobs.from(file);
+              if (!layersAreCompressed) {
+                Path compressedFile = destination.resolve(layerFiles.get(index) + ".compressed");
+                try (GZIPOutputStream compressorStream =
+                        new GZIPOutputStream(Files.newOutputStream(compressedFile));
+                    NotifyingOutputStream notifyingOutputStream =
+                        new NotifyingOutputStream(compressorStream, throttledProgressReporter)) {
+                  blob.writeTo(notifyingOutputStream);
+                }
+                blob = Blobs.from(compressedFile);
               }
-              blob = Blobs.from(compressedFile);
+
+              CachedLayer layer = cache.writeTarLayer(diffId, blob);
```

Comment:
Remember that this introduces another layer of unnecessary inefficiency that it basically copies an already existing file into the cache directory.

- If the layer is already compressed, hypothetically we could just move it. However, the layer file from `docker save` is (for now) under a temp directory, which may be on a filesystem different than the one for the cache directory, and moving a file may degrade into copying in many cases. When things happen concurrently this may be tricky to handle. Just keep this in mind for later optimization.

- If it is not compressed, I think this is an easy case to avoid copying. Instead of compressing the file into another file first and then let `writeTarLayer` to copy it, the blob can be constructed into the way that `writeTarLayer` will save the content directly while compressing. For example,

```java
Blob blobToBeCompressedWhenWritten =
    Blobs.from(outputStream -> // This will be the "cache file output stream" that writeTarLayer will open for writing.
        try (... = new GZIPOutputStream(outputStream); ...) {
            uncompressedLayerBlob.writeTo(notifyingOutputStream); })
```
Untested though.

Reply:
Actually, wouldn't concurrency matter in any case? Is this handled at a lower level inside `Cache`?

Reply:
Worth noting for the first point that we do need to write out the pre-compressed layers anyway, because we need to calculate the layer digest.

Reply:
Also, we should be fine as far as concurrency goes. As with the registry case, we write each local base image layer to a unique temporary directory first before moving them atomically to their final location in the cache.